python/ray/serve/autoscaling_metrics.py

```diff
@@ -24,39 +28,51 @@ def start_metrics_pusher(
         collection_callback: a callable that returns the metric data points to
           be sent to the the controller. The collection callback should take
           no argument and returns a dictionary of str_key -> float_value.
-        controller_handle: actor handle to Serve controller.
+        metrics_process_func: actor handle function.
+        stop_event: the daemon thread will be closed when the stop event is set
     """
 
     def send_once():
         data = collection_callback()
+
         # TODO(simon): maybe wait for ack or handle controller failure?
-        return controller_handle.record_autoscaling_metrics.remote(
-            data=data, send_timestamp=time.time()
-        )
+        return metrics_process_func.remote(data=data, send_timestamp=time.time())
 
-    def send_forever():
+    def send_forever(stop_event):
         last_ref: Optional[ray.ObjectRef] = None
         last_send_succeeded: bool = True
 
         while True:
             start = time.time()
-
-            if last_ref:
-                ready_refs, _ = ray.wait([last_ref], timeout=0)
-                last_send_succeeded = len(ready_refs) == 1
-            if last_send_succeeded:
-                last_ref = send_once()
+            if stop_event and stop_event.is_set():
+                return
+
+            if ray.is_initialized():
+                try:
+                    if last_ref:
+                        ready_refs, _ = ray.wait([last_ref], timeout=0)
+                        last_send_succeeded = len(ready_refs) == 1
+                    if last_send_succeeded:
+                        last_ref = send_once()
+                except Exception as e:
+                    logger.warning(
+                        "Autoscaling metrics pusher thread "
```

Comment:
I think we should elaborate on this error a bit. How about changing it to `"Autoscaling metrics-pusher thread failing to submit metrics to the controller, so [deployment_name] deployment cannot autoscale correctly. [Add info about what the user can do]: %s"`

If the deployment name can't be obtained here (or if it requires too much refactoring), we can omit it.

Reply:
+1